dallinger/experiment_server/sockets.py

```diff
@@ -20,107 +23,140 @@ def log(msg, level='info'):
-            log('Removed client {} from channel {}'.format(client, channel))
-
-    def send(self, client, data):
-        """Send data to one client.
-
-        Automatically discards invalid connections.
+        This is run continuously in a separate greenlet.
         """
+        pubsub = conn.pubsub()
         try:
-            client.send(data.decode('utf-8'))
-        except socket.error:
-            for channel in self.clients:
-                self.unsubscribe(client, channel)
-        else:
-            log('Sent to {}: {}'.format(client, data), level='debug')
-
-    def run(self):
-        """Listens for new messages in redis, and sends them to clients."""
-        log('Listening for messages')
-        for message in self.pubsub.listen():
+            pubsub.subscribe([self.name])
+        except ConnectionError:
+            app.logger.exception('Could not connect to redis.')
+        log('Listening on channel {}'.format(self.name))
+        for message in pubsub.listen():
             data = message.get('data')
             if message['type'] == 'message' and data != 'None':
                 channel = message['channel']
-                count = len(self.clients[channel])
-                if count:
-                    for client in self.clients[channel]:
-                        gevent.spawn(
-                            self.send, client, '{}:{}'.format(channel, data))
+                raw = '{}:{}'.format(channel, data)
+                for client in self.clients:
+                    gevent.spawn(client.send, raw)
+            gevent.sleep(0.001)
 
     def start(self):
-        """Starts listening in the background."""
-        self.greenlet = gevent.spawn(self.run)
+        """Start relaying messages."""
+        self.greenlet = gevent.spawn(self.listen)
 
     def stop(self):
-        if self.greenlet is not None:
+        """Stop relaying messages."""
+        if self.greenlet:
             self.greenlet.kill()
             self.greenlet = None
 
-    def heartbeat(self, ws):
-        """Send a ping to the websocket client periodically"""
-        while not ws.closed:
-            gevent.sleep(HEARTBEAT_DELAY)
-            gevent.spawn(self.send, ws, 'ping')
+
+class ChatBackend(object):
+    """Manages subscriptions of clients to multiple channels."""
+
+    def __init__(self):
+        self.channels = {}
+
+    def subscribe(self, client, channel_name):
+        """Register a new client to receive messages on a channel."""
+        if channel_name not in self.channels:
+            self.channels[channel_name] = channel = Channel(channel_name)
+            channel.start()
+
+        self.channels[channel_name].subscribe(client)
+
+    def unsubscribe(self, client):
+        """Unsubscribe a client from all channels."""
+        for channel in self.channels.values():
+            channel.unsubscribe(client)
 
 
+# There is one chat backend per process.
 chat_backend = ChatBackend()
 
 
+class Client(object):
```

Comment:
Glorious.

Reply:
I was particularly happy about the results of this refactoring.